private/bufpkg/bufmodule/bufmodulestore/module_data_store.go

```diff
@@ -167,14 +182,37 @@ func (p *moduleDataStore) getModuleDataForModuleKey(
 			return nil, err
 		}
 	} else {
-		moduleCacheBucket, err = p.getReadWriteBucketForDir(moduleKey)
-		// Not checking for fs.ErrNotExist. Function only returns error on actual system error.
+		dirPath, err := getModuleDataStoreDirPath(moduleKey)
 		if err != nil {
 			return nil, err
 		}
+		p.logDebugModuleKey(
+			moduleKey,
+			"module data store dir read write bucket",
+			zap.String("dirPath", dirPath),
+		)
+		moduleCacheBucket = storage.MapReadWriteBucket(p.bucket, storage.MapOnPrefix(dirPath))
+		// Only attempt to get a file lock when storing individual files
+		if p.filelocker != nil {
+			unlocker, err := p.filelocker.RLock(ctx, normalpath.Join(dirPath, externalModuleDataLockFileName))
+			if err != nil {
+				p.logger.Debug("failed_get_rlock", zap.Error(err))
+				return nil, err
+			}
+			defer func() {
+				if err := unlocker.Unlock(); err != nil {
+					p.logger.Debug("failed_get_runlock", zap.Error(err))
+					retErr = multierr.Append(retErr, err)
+				}
+			}()
+		}
 	}
+	// We want to avoid calling deleteInvalidModuleData if no module.yaml file was found,
+	// since it is an expensive operation and there is no valid module data, so the cache
+	// will be overwritten.
+	var externalModuleDataFileNotFound bool
 	defer func() {
-		if retErr != nil {
+		if retErr != nil && !externalModuleDataFileNotFound {
```

Comment:
I think we should rethink deleting module data at all for directory (not tar) module stores in the Get path. We are deciding to delete data while a read lock is held - by the time a write lock is acquired it is possible that the underlying data is now valid.

We probably want to have a test that covers this too - so far I think we only test writing invalid data with a storagemem backed bucket - we should update `module_data_store_test.go` to also exercise those tests against a storageos bucket w/ t.Tempdir() too.

Reply:
This makes sense to me -- we also cannot upgrade the lock, so grabbing the exclusive lock would require some messing around. Also, `GetModuleDatasForModuleKeys` is basically relying on `deleteInvalidModuleData` to return the `fs.NotExist` error, when in effect, we should be treating all errors in the "get" as a cache miss.